spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java

```diff
@@ -1085,4 +1088,43 @@ private ExecutorService getService() {
       return service;
     }
   }
+
+  /**
+   * Returns the first partition spec in an IcebergTable that shares the same names and ordering as
+   * the partition columns in a given Spark Table. Throws an error if not found
+   */
+  private static PartitionSpec findCompatibleSpec(
+      Table icebergTable, SparkSession spark, String sparkTable) throws AnalysisException {
+    List<String> parts = Lists.newArrayList(Splitter.on('.').limit(2).split(sparkTable));
+    String db = parts.size() == 1 ? "default" : parts.get(0);
+    String table = parts.get(parts.size() == 1 ? 0 : 1);
+
+    List<String> sparkPartNames =
+        spark.catalog().listColumns(db, table).collectAsList().stream()
+            .filter(org.apache.spark.sql.catalog.Column::isPartition)
+            .map(org.apache.spark.sql.catalog.Column::name)
+            .map(name -> name.toLowerCase(Locale.ROOT))
+            .collect(Collectors.toList());
+
+    for (PartitionSpec icebergSpec : icebergTable.specs().values()) {
```

Comment:
I see that the [specs are in an unordered map.](https://github.com/apache/iceberg/blob/25e7897b1161d5548ff428e0ec9016d5934635a1/core/src/main/java/org/apache/iceberg/util/PartitionUtil.java#L104) Should we iterate the specs from latest to oldest ? So that there is a predictable behavior in case the source spec can match multiple specs ? Or is it not a possible scenario?

Reply:
I don't think we have this in the spec, but the implementation here will always re-use a spec that is identical rather than creating a new identical spec. That said either would be valid if duplicates did exist but I don't think they should.